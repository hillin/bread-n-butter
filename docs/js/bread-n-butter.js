/**
 * Represents a parsing action; typically not created directly via `new`.
 */
export class Parser {
    /**
     * Creates a new custom parser that performs the given parsing action.
     */
    constructor(action) {
        this.action = action;
    }
    /**
     * Returns a parse result with either the value or error information.
     */
    parse(input) {
        const location = { index: 0, line: 1, column: 1 };
        const context = new Context(input, location);
        const result = this.and(eof).action(context);
        if (result.type === "ActionOK") {
            return {
                type: "ParseOK",
                value: result.value[0],
            };
        }
        return {
            type: "ParseFail",
            location: result.furthest,
            expected: result.expected,
        };
    }
    /**
     * Returns the parsed result or throws an error.
     */
    tryParse(input) {
        const result = this.parse(input);
        if (result.type === "ParseOK") {
            return result.value;
        }
        const { expected, location } = result;
        const { line, column } = location;
        const message = `parse error at line ${line} column ${column}: ` +
            `expected ${expected.join(", ")}`;
        throw new Error(message);
    }
    /**
     * Combines two parsers one after the other, yielding the results of both in
     * an array.
     */
    and(parserB) {
        return new Parser((context) => {
            const a = this.action(context);
            if (a.type === "ActionFail") {
                return a;
            }
            context = context.moveTo(a.location);
            const b = context.merge(a, parserB.action(context));
            if (b.type === "ActionOK") {
                const value = [a.value, b.value];
                return context.merge(b, context.ok(b.location.index, value));
            }
            return b;
        });
    }
    /**
     * Try to parse using the current parser. If that fails, parse using the
     * second parser.
     */
    or(parserB) {
        return new Parser((context) => {
            const a = this.action(context);
            if (a.type === "ActionOK") {
                return a;
            }
            return context.merge(a, parserB.action(context));
        });
    }
    /**
     * Parse using the current parser. If it succeeds, pass the value to the
     * callback function, which returns the next parser to use.
     */
    chain(fn) {
        return new Parser((context) => {
            const a = this.action(context);
            if (a.type === "ActionFail") {
                return a;
            }
            const parserB = fn(a.value);
            context = context.moveTo(a.location);
            return context.merge(a, parserB.action(context));
        });
    }
    /**
     * Yields the value from the parser after being called with the callback.
     */
    map(fn) {
        return this.chain((a) => {
            return ok(fn(a));
        });
    }
    /**
     * Returns the callback called with the parser.
     */
    thru(fn) {
        return fn(this);
    }
    /**
     * Returns a parser which parses the same value, but discards other error
     * messages, using the ones supplied instead.
     */
    desc(expected) {
        return new Parser((context) => {
            const result = this.action(context);
            if (result.type === "ActionOK") {
                return result;
            }
            return { type: "ActionFail", furthest: result.furthest, expected };
        });
    }
    /**
     * Wraps the current parser with before & after parsers.
     */
    wrap(before, after) {
        return before.and(this).chain(([, value]) => {
            return after.map(() => value);
        });
    }
    /**
     * Ignores content before and after the current parser, based on the supplied
     * parser.
     */
    trim(beforeAndAfter) {
        return this.wrap(beforeAndAfter, beforeAndAfter);
    }
    /**
     * Repeats the current parser zero or more times, yielding the results in an
     * array.
     */
    many0() {
        return this.many1().or(ok([]));
    }
    /**
     * Parsers the current parser **one** or more times. See `many0` for more
     * details.
     */
    many1() {
        return new Parser((context) => {
            const items = [];
            let result = this.action(context);
            if (result.type === "ActionFail") {
                return result;
            }
            while (result.type === "ActionOK") {
                items.push(result.value);
                if (result.location.index === context.location.index) {
                    throw new Error("infinite loop detected; don't call many0 or many1 with parsers that can accept zero characters");
                }
                context = context.moveTo(result.location);
                result = context.merge(result, this.action(context));
            }
            return context.merge(result, context.ok(context.location.index, items));
        });
    }
    /**
     * Returns a parser that parses zero or more times, separated by the separator
     * parser supplied.
     */
    sepBy0(separator) {
        return this.sepBy1(separator).or(ok([]));
    }
    /**
     * Returns a parser that parses one or more times, separated by the separator
     * parser supplied.
     */
    sepBy1(separator) {
        return this.chain((first) => {
            return separator
                .and(this)
                .map(([, value]) => value)
                .many0()
                .map((rest) => {
                return [first, ...rest];
            });
        });
    }
    /**
     * Returns a parser that adds name and start/end location metadata.
     */
    node(name) {
        return location.and(this).chain(([start, value]) => {
            return location.map((end) => {
                const type = "ParseNode";
                return { type, name, value, start, end };
            });
        });
    }
}
/**
 * Parser that yields the current `SourceLocation`, containing properties
 * `index`, `line` and `column`.
 */
export const location = new Parser((context) => {
    return context.ok(context.location.index, context.location);
});
/**
 * Returns a parser that yields the given value and consumes no input.
 */
export function ok(value) {
    return new Parser((context) => {
        return context.ok(context.location.index, value);
    });
}
/**
 * Returns a parser that fails with the given messages and consumes no input.
 */
export function fail(expected) {
    return new Parser((context) => {
        return context.fail(context.location.index, expected);
    });
}
/**
 * This parser succeeds if the input has already been fully parsed.
 */
export const eof = new Parser((context) => {
    if (context.location.index < context.input.length) {
        return context.fail(context.location.index, ["<EOF>"]);
    }
    return context.ok(context.location.index, "<EOF>");
});
/** Returns a parser that matches the exact text supplied. */
export function text(string) {
    return new Parser((context) => {
        const start = context.location.index;
        const end = start + string.length;
        if (context.input.slice(start, end) === string) {
            return context.ok(end, string);
        }
        return context.fail(start, [string]);
    });
}
/**
 * Returns a parser that matches the entire regular expression at the current
 * parser position.
 */
export function match(regexp) {
    for (const flag of regexp.flags) {
        switch (flag) {
            case "i": // ignoreCase
            case "s": // dotAll
            case "m": // multiline
            case "u": // unicode
                continue;
            default:
                throw new Error("only the regexp flags 'imsu' are supported");
        }
    }
    const sticky = new RegExp(regexp.source, regexp.flags + "y");
    return new Parser((context) => {
        const start = context.location.index;
        sticky.lastIndex = start;
        const match = context.input.match(sticky);
        if (match) {
            const end = start + match[0].length;
            const string = context.input.slice(start, end);
            return context.ok(end, string);
        }
        return context.fail(start, [String(regexp)]);
    });
}
/**
 * Takes a lazily invoked callback that returns a parser, so you can create
 * recursive parsers.
 */
export function lazy(fn) {
    // NOTE: This parsing action overwrites itself on the specified parser. We're
    // assuming that the same parser won't be returned to multiple `lazy` calls. I
    // never heard of such a thing happening in Parsimmon, and it doesn't seem
    // likely to happen here either. I assume this is faster than using variable
    // closure and an `if`-statement here, but I honestly don't know.
    const parser = new Parser((context) => {
        parser.action = fn().action;
        return parser.action(context);
    });
    return parser;
}
function union(a, b) {
    return [...new Set([...a, ...b])];
}
/**
 * Represents the current parsing context.
 */
class Context {
    constructor(input, location) {
        this.input = input;
        this.location = location;
    }
    /**
     * Returns a new context with the supplied location and the current input.
     */
    moveTo(location) {
        return new Context(this.input, location);
    }
    _internal_move(index) {
        if (index === this.location.index) {
            return this.location;
        }
        const start = this.location.index;
        const end = index;
        const chunk = this.input.slice(start, end);
        let { line, column } = this.location;
        for (const ch of chunk) {
            if (ch === "\n") {
                line++;
                column = 1;
            }
            else {
                column++;
            }
        }
        return { index, line, column };
    }
    /**
     * Represents a successful parse ending before the given `index`, with the
     * specified `value`.
     */
    ok(index, value) {
        return {
            type: "ActionOK",
            value,
            location: this._internal_move(index),
            furthest: { index: -1, line: -1, column: -1 },
            expected: [],
        };
    }
    /**
     * Represents a failed parse starting at the given `index`, with the specified
     * list `expected` messages (note: this list usually only has one item).
     */
    fail(index, expected) {
        return {
            type: "ActionFail",
            furthest: this._internal_move(index),
            expected,
        };
    }
    /**
     * Merge two sequential `ActionResult`s so that the `expected` and location data
     * is preserved correctly.
     */
    merge(a, b) {
        if (b.furthest.index > a.furthest.index) {
            return b;
        }
        const expected = b.furthest.index === a.furthest.index
            ? union(a.expected, b.expected)
            : a.expected;
        if (b.type === "ActionOK") {
            return {
                type: "ActionOK",
                location: b.location,
                value: b.value,
                furthest: a.furthest,
                expected,
            };
        }
        return {
            type: "ActionFail",
            furthest: a.furthest,
            expected,
        };
    }
}
