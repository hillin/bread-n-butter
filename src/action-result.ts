import { SourceLocation } from "./source-location";

export type ActionResult<A> = ActionOK<A> | ActionFail;

export class ActionOK<A> {
  location: SourceLocation;
  value: A;
  furthest: SourceLocation;
  expected: string[];

  constructor(
    location: SourceLocation,
    value: A,
    furthest: SourceLocation,
    expected: string[]
  ) {
    this.location = location;
    this.value = value;
    this.furthest = furthest;
    this.expected = expected;
  }

  isOK(): this is ActionOK<A> {
    return true;
  }

  merge<B>(b: ActionResult<B>): ActionResult<B> {
    return merge(b, this);
  }
}

export class ActionFail {
  furthest: SourceLocation;
  expected: string[];

  constructor(furthest: SourceLocation, expected: string[]) {
    this.furthest = furthest;
    this.expected = expected;
  }

  isOK<A>(): this is ActionOK<A> {
    return false;
  }

  merge<B>(b: ActionResult<B>): ActionResult<B> {
    return merge(b, this);
  }
}

function merge<A, B>(a: ActionResult<A>, b: ActionResult<B>): ActionResult<A> {
  if (a.furthest.index > b.furthest.index) {
    return a;
  }
  const expected =
    a.furthest.index === b.furthest.index
      ? union(a.expected, b.expected)
      : b.expected;
  if (a.isOK()) {
    return new ActionOK(a.location, a.value, b.furthest, expected);
  }
  return new ActionFail(b.furthest, expected);
}

function union(a: string[], b: string[]): string[] {
  return [...new Set([...a, ...b])].sort();
}
