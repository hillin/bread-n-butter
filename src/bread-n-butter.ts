class Parser<Value> {
  readonly action: (context: Context) => Result<Value>;

  constructor(action: (context: Context) => Result<Value>) {
    this.action = action;
  }

  parse(input: string): Result<Value> {
    const context = new Context({ input, index: 0 });
    return andThen(this, matchEOF)
      .action(context)
      .map((values) => values[0]);
  }
}

class Context {
  readonly input: string;
  readonly index: number;

  constructor(options: { input: string; index: number }) {
    this.input = options.input;
    this.index = options.index;
  }

  consume(amount: number): Context {
    return new Context({
      input: this.input,
      index: this.index + amount,
    });
  }
}

interface Result<Value> {
  readonly context: Context;
  map<NewValue>(
    fn: (value: Value, context: Context) => NewValue
  ): Result<NewValue>;
  flatMap<NewValue>(
    fn: (value: Value, context: Context) => Result<NewValue>
  ): Result<NewValue>;
}

class OK<Value> {
  readonly value: Value;
  readonly context: Context;

  constructor(options: { value: Value; context: Context }) {
    this.value = options.value;
    this.context = options.context;
  }

  map<NewValue>(
    fn: (value: Value, context: Context) => NewValue
  ): Result<NewValue> {
    return new OK({
      value: fn(this.value, this.context),
      context: this.context,
    });
  }

  flatMap<NewValue>(
    fn: (value: Value, context: Context) => Result<NewValue>
  ): Result<NewValue> {
    return fn(this.value, this.context);
  }
}

class Fail<Value> implements Result<Value> {
  readonly messages: readonly string[];
  readonly context: Context;

  constructor(options: { messages: readonly string[]; context: Context }) {
    this.messages = [...options.messages];
    this.context = options.context;
  }

  map<NewValue>(
    _fn: (value: Value, context: Context) => NewValue
  ): Result<NewValue> {
    return new Fail({ messages: this.messages, context: this.context });
  }

  flatMap<NewValue>(
    _fn: (value: Value, context: Context) => Result<NewValue>
  ): Result<NewValue> {
    return new Fail({ messages: this.messages, context: this.context });
  }
}

export function custom<Value>(
  action: (context: Context) => Result<Value>
): Parser<Value> {
  return new Parser(action);
}

export function ok<Value>(value: Value, context: Context): Result<Value> {
  return new OK({ value, context });
}

export function fail<Value>(message: string, context: Context): Result<Value> {
  return new Fail({ messages: [message], context });
}

export function of<Value>(value: Value): Parser<Value> {
  return custom((context) => ok(value, context));
}

export function andThen<Value1, Value2>(
  parser1: Parser<Value1>,
  parser2: Parser<Value2>
): Parser<readonly [Value1, Value2]> {
  return custom((context) => {
    return parser1.action(context).flatMap((value1, context) => {
      return parser2.action(context).map((value2, _context) => {
        return [value1, value2] as const;
      });
    });
  });
}

export function matchString(string: string): Parser<string> {
  return custom((context) => {
    const chunk = context.input.slice(
      context.index,
      context.index + string.length
    );
    if (chunk === string) {
      return ok(string, context.consume(1));
    }
    return fail(string, context);
  });
}

export const matchEOF = custom((context) => {
  if (context.index >= context.input.length) {
    return ok("end of file", context);
  }
  return fail("end of file", context);
});

export function matchRegExp(regexp: RegExp): Parser<string> {
  return custom((context) => {
    const stickyRegexp = new RegExp(regexp, "y");
    stickyRegexp.lastIndex = context.index;
    const match = context.input.match(stickyRegexp);
    if (!match) {
      return fail(regexp.toString(), context);
    }
    return ok(match[0], context.consume(match[0].length));
  });
}
