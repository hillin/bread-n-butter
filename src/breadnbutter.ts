interface Parser<Value> {
  (context: ParserContext): Result<Value>;
}

interface ParserContext {
  type: "bnb.ParserContext";
  input: string;
  index: number;
}

interface OK<Value> {
  type: "bnb.OK";
  value: Value;
  context: ParserContext;
}

interface Fail {
  type: "bnb.Fail";
  message: string;
  context: ParserContext;
}

type Result<Value> = OK<Value> | Fail;

export function custom<Value>(
  handler: (context: ParserContext) => Result<Value>
): Parser<Value> {
  return handler;
}

export function ok<Value>(value: Value, context: ParserContext): OK<Value> {
  return {
    type: "bnb.OK",
    value,
    context,
  };
}

export function fail(message: string, context: ParserContext): Fail {
  return {
    type: "bnb.Fail",
    message,
    context,
  };
}

export function parse<Value>(
  parser: Parser<Value>,
  input: string
): Result<Value> {
  const context: ParserContext = {
    type: "bnb.ParserContext",
    input,
    index: 0,
  };
  const result = andThen(parser, matchEOF)(context);
  if (result.type === "bnb.OK") {
    const [value] = result.value;
    return ok(value, result.context);
  }
  return result;
}

export function of<Value>(value: Value): Parser<Value> {
  return custom((context) => ok(value, context));
}

export function andThen<Value1, Value2>(
  parser1: Parser<Value1>,
  parser2: Parser<Value2>
): Parser<readonly [Value1, Value2]> {
  return custom((context) => {
    const result1 = parser1(context);
    if (result1.type === "bnb.Fail") {
      return result1;
    }
    const result2 = parser2(result1.context);
    if (result2.type === "bnb.Fail") {
      return result2;
    }
    const value = [result1.value, result2.value] as const;
    return ok(value, result2.context);
  });
}

export function matchString(string: string): Parser<string> {
  return custom((context) => {
    const chunk = context.input.slice(
      context.index,
      context.index + string.length
    );
    if (chunk === string) {
      return ok(string, {
        ...context,
        index: context.index + 1,
      });
    }
    return fail(string, context);
  });
}

export const matchEOF = custom((context) => {
  if (context.index === context.input.length) {
    return ok("end of file", context);
  }
  return fail("end of file", context);
});

export function matchRegExp(regexp: RegExp): Parser<string> {
  return custom((context) => {
    const stickyRegexp = new RegExp(regexp, "y");
    stickyRegexp.lastIndex = context.index;
    const match = context.input.match(stickyRegexp);
    console.log("MATCH", match);
    if (!match) {
      return fail(regexp.toString(), context);
    }
    return ok(match[0], {
      ...context,
      index: context.index + match[0].length,
    });
  });
}
