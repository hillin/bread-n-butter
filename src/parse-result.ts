import { SourceLocation } from "./source-location";

export type ParseResult<A> = ParseOK<A> | ParseFail;

export class ParseOK<A> {
  readonly value: A;

  constructor(value: A) {
    this.value = value;
  }

  isOK(): this is ParseOK<A> {
    return true;
  }
}

export class ParseFail {
  // TODO: Possibly yoink the properties off location and put them straight on
  // this class so it's easier to look at the object?
  readonly location: SourceLocation;
  readonly expected: readonly string[];

  constructor(location: SourceLocation, expected: readonly string[]) {
    this.location = location;
    this.expected = expected;
  }

  isOK<A>(): this is ParseOK<A> {
    return false;
  }
}
