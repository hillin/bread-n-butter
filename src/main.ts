import * as bnb from "./bread-n-butter";

const a = bnb.str("a");
const b = bnb.str("b");
const ab = a.and(b);
const aOrB = a.or(b);
const abOrA = ab.or(a);
const balance = bnb
  .str("(")
  .or(bnb.str("["))
  .chain((value) => {
    if (value === "(") {
      return bnb.str(")");
    } else {
      return bnb.str("]");
    }
  });
const justX = bnb.of("X");
// TODO: Are these answers correct?
console.log(aOrB.parse("a"));
console.log(aOrB.parse("b"));
console.log(aOrB.parse("ab"));
console.log(aOrB.parse(""));
console.log(abOrA.parse("ab"));
console.log(abOrA.parse("aa"));
console.log(abOrA.parse("a"));
console.log(abOrA.parse(""));
console.log(abOrA.parse("aba"));
console.log(balance.parse("()"));
console.log(balance.parse("[]"));
console.log(balance.parse("[)"));
console.log(balance.parse("["));
console.log(justX.parse(""));
