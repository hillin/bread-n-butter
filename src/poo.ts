import * as bnb from "./bread-n-butter";
