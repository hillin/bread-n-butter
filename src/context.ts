import { SourceLocation } from "./source-location";
import { ActionResult, ActionOK, ActionFail } from "./action-result";

export class Context {
  readonly input: string;
  readonly location: SourceLocation;

  constructor(input: string, location: SourceLocation) {
    this.input = input;
    this.location = location;
  }

  withLocation(location: SourceLocation): Context {
    return new Context(this.input, location);
  }

  move(index: number): SourceLocation {
    const start = this.location.index;
    const end = index;
    const chunk = this.input.slice(start, end);
    return this.location.add(chunk);
  }

  ok<A>(index: number, value: A): ActionResult<A> {
    return new ActionOK(
      this.move(index),
      value,
      new SourceLocation(-1, -1, -1),
      []
    );
  }

  fail<A>(index: number, expected: readonly string[]): ActionResult<A> {
    return new ActionFail(this.move(index), expected);
  }
}
