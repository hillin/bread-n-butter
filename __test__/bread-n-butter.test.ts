import * as bnb from "../src/bread-n-butter";

test("test a test", () => {
  const isX = bnb.str("x");
  expect(isX.parse("x")).toMatchSnapshot();
});

test("and", () => {
  const isX = bnb.str("x");
  const isY = bnb.str("y");
  const isXY = isX.and(isY);
  expect(isXY.parse("xy")).toMatchSnapshot();
});

test("and triple", () => {
  const isX = bnb.str("x");
  const isY = bnb.str("y");
  const isZ = bnb.str("z");
  const isXYZ = isX.and(isY).and(isZ);
  expect(isXYZ.parse("xyz")).toMatchSnapshot();
});

test("and success triple 2", () => {
  const isX = bnb.str("x");
  const isY = bnb.str("y");
  const isZ = bnb.str("z");
  const isXYZ = isX.and(isY).chain(([x, y]) => {
    return isZ.map((z) => {
      return [x, y, z] as const;
    });
  });
  expect(isXYZ.parse("xyz")).toMatchSnapshot();
});

test("and failure", () => {
  const isX = bnb.str("x");
  const isY = bnb.str("y");
  const isXY = isX.and(isY);
  expect(isXY.parse("x")).toMatchSnapshot();
  expect(isXY.parse("y")).toMatchSnapshot();
  expect(isXY.parse("yx")).toMatchSnapshot();
  expect(isXY.parse("")).toMatchSnapshot();
});

test("sepBy0", () => {
  const isA = bnb.str("a");
  const isSep = bnb.str(",");
  const isAList = isA.sepBy0(isSep);
  expect(isAList.parse("")).toMatchSnapshot();
  expect(isAList.parse("a")).toMatchSnapshot();
  expect(isAList.parse("a,a")).toMatchSnapshot();
  expect(isAList.parse("a,a,a")).toMatchSnapshot();
  expect(isAList.parse("a,a,b")).toMatchSnapshot();
  expect(isAList.parse("b")).toMatchSnapshot();
});

test("sepBy1", () => {
  const isA = bnb.str("a");
  const isSep = bnb.str(",");
  const isAList = isA.sepBy1(isSep);
  expect(isAList.parse("")).toMatchSnapshot();
  expect(isAList.parse("a")).toMatchSnapshot();
  expect(isAList.parse("a,a")).toMatchSnapshot();
  expect(isAList.parse("a,a,a")).toMatchSnapshot();
  expect(isAList.parse("a,a,b")).toMatchSnapshot();
  expect(isAList.parse("b")).toMatchSnapshot();
});

test("many0", () => {
  const isA = bnb.str("a");
  const isAAA = isA.many0();
  expect(isAAA.parse("")).toMatchSnapshot();
  expect(isAAA.parse("a")).toMatchSnapshot();
  expect(isAAA.parse("aa")).toMatchSnapshot();
  expect(isAAA.parse("aaa")).toMatchSnapshot();
  expect(isAAA.parse("aaaa")).toMatchSnapshot();
  expect(isAAA.parse("b")).toMatchSnapshot();
});

test("many1", () => {
  const isA = bnb.str("a");
  const isAAA = isA.many1();
  expect(isAAA.parse("a")).toMatchSnapshot();
  expect(isAAA.parse("aa")).toMatchSnapshot();
  expect(isAAA.parse("aaa")).toMatchSnapshot();
  expect(isAAA.parse("aaaa")).toMatchSnapshot();
  expect(isAAA.parse("")).toMatchSnapshot();
  expect(isAAA.parse("b")).toMatchSnapshot();
});

test("or", () => {
  const isA = bnb.str("a");
  const isB = bnb.str("b");
  const isAorB = isA.or(isB);
  expect(isAorB.parse("a")).toMatchSnapshot();
  expect(isAorB.parse("b")).toMatchSnapshot();
  expect(isAorB.parse("c")).toMatchSnapshot();
  expect(isAorB.parse("ab")).toMatchSnapshot();
  expect(isAorB.parse("")).toMatchSnapshot();
});

test("matchEOF", () => {
  const isEOF = bnb.eof;
  expect(isEOF.parse("")).toMatchSnapshot();
  expect(isEOF.parse("x")).toMatchSnapshot();
});

test("match", () => {
  const isNumber = bnb.match(/\d+/);
  expect(isNumber.parse("12")).toMatchSnapshot();
  expect(isNumber.parse("abc")).toMatchSnapshot();
});

test("lisp lists", () => {
  const isSymbol = bnb.match(/[a-zA-Z_-]+/);
  const isLP = bnb.str("(");
  const isRP = bnb.str(")");
  const isWS = bnb.match(/\s+/);
  const isList = isLP.chain(() => {
    return isSymbol.sepBy0(isWS).chain((values) => {
      return isRP.map(() => {
        return values;
      });
    });
  });
  expect(isList.parse("(a b c)")).toMatchSnapshot();
});

test("node", () => {
  const identifier = bnb
    .match(/[a-z]+/i)
    .node("Identifier")
    .desc("identifier");
  expect(identifier.parse("abc")).toMatchSnapshot();
});
