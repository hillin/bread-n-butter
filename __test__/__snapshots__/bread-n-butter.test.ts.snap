// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`andThen failure 1 1`] = `
Fail {
  "context": Context {
    "index": 1,
    "input": "x",
  },
  "messages": Array [
    "y",
  ],
}
`;

exports[`andThen failure 2 1`] = `
Fail {
  "context": Context {
    "index": 0,
    "input": "y",
  },
  "messages": Array [
    "x",
  ],
}
`;

exports[`andThen failure 3 1`] = `
Fail {
  "context": Context {
    "index": 0,
    "input": "yx",
  },
  "messages": Array [
    "x",
  ],
}
`;

exports[`andThen failure 4 1`] = `
Fail {
  "context": Context {
    "index": 0,
    "input": "",
  },
  "messages": Array [
    "x",
  ],
}
`;

exports[`andThen success 1`] = `
OK {
  "context": Context {
    "index": 2,
    "input": "xy",
  },
  "value": Array [
    "x",
    "y",
  ],
}
`;

exports[`matchEOF failure 1`] = `
Fail {
  "context": Context {
    "index": 0,
    "input": "x",
  },
  "messages": Array [
    "end of file",
  ],
}
`;

exports[`matchEOF success 1`] = `
OK {
  "context": Context {
    "index": 0,
    "input": "",
  },
  "value": "end of file",
}
`;

exports[`matchRegExp failure 1`] = `
Fail {
  "context": Context {
    "index": 0,
    "input": "abc",
  },
  "messages": Array [
    "/\\\\d+/",
  ],
}
`;

exports[`matchRegExp success 1`] = `
OK {
  "context": Context {
    "index": 2,
    "input": "12",
  },
  "value": "12",
}
`;

exports[`test a test 1`] = `
OK {
  "context": Context {
    "index": 1,
    "input": "x",
  },
  "value": "x",
}
`;
